frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,368 @@
+  }
+  return newMessage
+}
+
+export function getLatestMessages ({
+  count, messages
+}: { count: number, messages: Array<Object> }): Array<Object> {
+  return messages.slice(Math.max(messages.length - count, 0))
+}
+
+export async function leaveChatRoom ({ contractID }: {
+  contractID: string
+}) {
+  const rootState = sbp('state/vuex/state')
+  if (contractID === rootState.currentChatRoomId) {
+    await sbp('state/vuex/commit', 'setCurrentChatRoomId', {
+      groupId: rootState.currentGroupId
+    })
+    const curRouteName = sbp('controller/router').history.current.name
+    if (curRouteName === 'GroupChat' || curRouteName === 'GroupChatConversation') {
+      sbp('controller/router').push({ name: 'GroupChat' })
+    }
+  }
+  sbp('state/vuex/commit', 'removeContract', contractID)
+}
+
+function createNotificationData (
+  notificationType: string,
+  channelName: string,
+  moreParams: Object = {}
+): Object {
+  return {
+    type: MESSAGE_TYPES.NOTIFICATION,
+    notification: {
+      type: notificationType,
+      channelName,
+      ...moreParams
+    }
+  }
+}
+
+function emitMessageEvents ({ type, contractID, hash, state }: {
+  type: string, contractID: string, hash: string, state: Object
+}): void {
+  for (let i = state.messages.length - 1; i >= 0; i--) {
+    if (state.messages[i].id === hash) {
+      sbp('okTurtles.events/emit', `${CHATROOM_MESSAGE_ACTION}-${contractID}`, {
+        type,
+        data: { message: state.messages[i] }
+      })
+      break
+    }
+  }
+}
+
+sbp('chelonia/defineContract', {
+  name: 'gi.contracts/chatroom',
+  metadata: {
+    validate: objectOf({
+      createdDate: string, // action created date
+      username: string, // action creator
+      identityContractID: string // action creator identityContractID
+    }),
+    create () {
+      const { username, identityContractID } = sbp('state/vuex/state').loggedIn
+      return {
+        createdDate: new Date().toISOString(),
+        username,
+        identityContractID
+      }
+    }
+  },
+  state (contractID) {
+    return sbp('state/vuex/state')[contractID]
+  },
+  getters: {
+    currentChatRoomState (state) {
+      return state
+    },
+    chatRoomSettings (state, getters) {
+      return getters.currentChatRoomState.settings || {}
+    },
+    chatRoomAttributes (state, getters) {
+      return getters.currentChatRoomState.attributes || {}
+    },
+    chatRoomUsers (state, getters) {
+      return getters.currentChatRoomState.users || {}
+    },
+    chatRoomLatestMessages (state, getters) {
+      const messages = getters.currentChatRoomState.messages || []
+      return getLatestMessages({
+        count: getters.chatRoomSettings.messagesPerPage,
+        messages
+      })
+    }
+  },
+  actions: {
+    // This is the constructor of Chat contract
+    'gi.contracts/chatroom': {
+      validate: chatRoomType,
+      process ({ meta, data }, { state }) {
+        const initialState = merge({
```

Comment:
Sorry, @taoeffect. I can not understand what you mean. I can see that the `merge` function inside `giLodash.js` is just to create a new object having the same values in depth. It uses `cloneDeep` which create a new object having the same value.

Reply:
@Silver-IT I'm saying that we should design our contracts to be cloneable.

This means being able to take the entire state of one contract, and plug it in to the constructor of another contract.

Look at how the group contract is implemented. It puts `data` at the very end of the `merge` function as the last parameter, by itself.

Instead, here you have `...data` being inserted into `attributes`. That doesn't allow cloning, since data could potentially contain everything outside here, like `settings, users, messages`, and `attributes`.

Reply:
Actually, I can understand what to do. But I don't understand why.
Why do the contracts need to be clonable?
Why it isn't clonable once I use this way above even if data could potentially ...?

Reply:
@Silver-IT Those are good questions.

> Why do the contracts need to be clonable?
> When and how do we clone the contracts?

Imagine for example you're moving to a different server and want to copy and re-create all of your contracts there. By making the constructor accept the entire state, it makes it easier to do this. You simply create a contract as normal, and pass in the entire state of the previous one, and boom, you have a cloned contract on the new server that you can use.

>  Why it isn't clonable once I use this way above even if data could potentially ...?

It's true, we could theoretically add a separate copying/cloning action that takes that data and replaces the state with it. So this is more a convenience convention than a necessity.

Are you against the modification? If so why?

Reply:
> Imagine for example you're moving to a different server and want to copy and re-create all of your contracts there.

Do you mean we don't run all the actions for all the contracts from the beginning and from the empty state and just clone the entire state at once and set the latest HEAD?

> It's true, we could theoretically add a separate copying/cloning action that takes that data and replaces the state with it. So this is more a convenience convention than a necessity.

Now, I understood why we need to do that. The state needs to be clonable if we send the previous state through the `data` variable, no matter which fields are there inside.

> Are you against the modification?

I wasn't against the modification. I wanted to know the reason why we need to keep it this way.

Reply:
> Do you mean we don't run all the actions for all the contracts from the beginning and from the empty state and just clone the entire state at once and set the latest HEAD?

Right, copying the latest built-up state for a specific contract and re-creating in another identical contract.

Reply:
That's awesome...